### DISCUSSION TITLE
WebTransport and RTMFP

### DISCUSSION THREAD

**Discussion by ROBERT-MCDOWELL** [zenomt/rtmfp-cpp#22] [Ideas]
Hi Michael,
do you think it's technically possible to write the RTMFP protocol on top of webtransport?

**Comment by zenomt** [comment]
yes, it would be technically possible to adapt _most_ of RTMFP to WebTransport. of course its P2P functions won't work that way, and there'd be little utility to RTMFP's startup handshake if a WebTransport session/connection is already established. and here i'm assuming you mean "WebTransport (QUIC)". and to get RTMFP's benefits, you'd probably want to use QUIC datagrams, not streams. and then RTMFP's congestion control would be fighting with QUIC's congestion control.

if one were going to use TCP and target web browsers, i'd suggest using [RTWebSocket](https://github.com/zenomt/rtwebsocket) instead, which gives you RTMFP's _flow semantics_ but over a plain old websocket that any browser can already talk. note that `tcserver` supports RTWebSocket today.

**Comment by ROBERT-MCDOWELL** [comment]
I'm following webtransport repo (HTTP3) since 4 years now and what I can now is congestion can be dsiabled, as well as p2p can be done with it. it's still in active development, things can still change.
If I understand well what you said, RTWebSocket use TCP to passthru a RTMFP datagram for data only right? no way to pass any stream or else?

**Reply by zenomt** [reply]
RTWebSocket is a distinct protocol on top of WebSocket messages that gives you the high-level abstraction of RTMFP: multiplexing multiple message-oriented flows, where flows can have different priorities, and each message can have a transmission deadline and can be abandoned even if transmission has started. each flow also has metadata, just like an RTMFP flow. for streaming, exactly the same mapping that's defined in RFC 7425 is used to send RTMP messages over RTMFP and RTWebSocket.

**Reply by ROBERT-MCDOWELL** [reply]
ah Ok understood, so it's only for message flow, not possible for audio/video streaming. and not possible for p2p too :(

**Reply by zenomt** [reply]
no, RTWebSocket is absolutely for audio/video streaming. of course a WebSocket is a client-server connection, so P2P is not possible.

**Reply by zenomt** [reply]
as i said, `tcserver` speaks RTMP-over-RTWebSocket, and the rtwebsocket repo has a `play-demo.html` page and supporting javascript that can connect to `tcserver` and play video/audio streams (in Chrome, using WebCodecs and WebAudio).

**Reply by zenomt** [reply]
also remember, the "F" in RTMFP stands for "Flow". flows are how all data is transported in RTMFP. i encourage you to read [RFC 7016](https://www.rfc-editor.org/rfc/rfc7016.html) (RTMFP) and at least  [Section 5 of RFC 7425](https://www.rfc-editor.org/rfc/rfc7425.html#section-5) (RTMP-over-RTMFP).

**Reply by ROBERT-MCDOWELL** [reply]
oops ok thanks Michael, I read this RFC long time ago (2010~) forgot completely  :). This is a very good repo you started!
